crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2309,102 +2318,80 @@ impl<'db> Class<'db> {
     /// Return the metaclass of this class, or `Unknown` if the metaclass cannot be inferred.
     pub(crate) fn metaclass(self, db: &'db dyn Db) -> Type<'db> {
         // TODO: `type[Unknown]` would be a more precise fallback
-        // (needs support for <https://docs.python.org/3/library/typing.html#the-type-of-class-objects>)
         self.try_metaclass(db).unwrap_or(Type::Unknown)
     }
 
     /// Return the metaclass of this class, or an error if the metaclass cannot be inferred.
     #[salsa::tracked]
     pub(crate) fn try_metaclass(self, db: &'db dyn Db) -> Result<Type<'db>, MetaclassError<'db>> {
-        /// Infer the metaclass of a class, tracking the classes that have been visited to detect
-        /// cyclic definitions.
-        fn infer<'db>(
-            db: &'db dyn Db,
-            class: Class<'db>,
-            seen: &mut SeenSet<Class<'db>>,
-        ) -> Result<Type<'db>, MetaclassError<'db>> {
-            // Recursively infer the metaclass of a class, ensuring that cyclic definitions are
-            // detected.
-            let mut safe_recurse = |class: Class<'db>| -> Result<Type<'db>, MetaclassError<'db>> {
-                // Each base must be considered in isolation.
-                let num_seen = seen.len();
-                if !seen.insert(class) {
-                    return Err(MetaclassError {
-                        kind: MetaclassErrorKind::CyclicDefinition,
-                    });
-                }
-                let metaclass = infer(db, class, seen)?;
-                seen.truncate(num_seen);
-                Ok(metaclass)
-            };
+        // Identify the class's own metaclass (or take the first base class's metaclass).
+        let mut base_classes = self.fully_static_explicit_bases(db).peekable();
 
-            let mut base_classes = class
-                .explicit_bases(db)
-                .iter()
-                .copied()
-                .filter_map(Type::into_class_literal);
+        if base_classes.peek().is_some() && self.is_cyclically_defined(db) {
+            // We emit diagnostics for cyclic class definitions elsewhere.
+            // Don't even try to infer the metaclass if the class is cyclically defined:
+            // it's impossible to do so without entering an infinite loop.
```

Comment:
nit: obviously it's not impossible, we did it prior to this diff :) might be clearer to just say "we'd enter an infinite loop."

Reply:
I mean... kinda? We would *try* to infer the metaclass prior to this diff, but we would immediately throw up our hands and bail out of it as soon as we detected a cycle. That feels like kind of the same thing to me 😄 but, I'll rewrite the comment :-)

Reply:
Ok, fair enough!